### PR TITLE
feat(examples): add image classification example using quantized vit …

### PR DESCRIPTION
**Pull request by shovan777**
# What does this PR do?
This adds an image classification example using a quantized ViT model as per #169.

## Before submitting
- [x] Did you read the [contributor guideline](https://github.com/huggingface/optimum-quanto/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [x] Was this discussed/approved via a Github issue or the [github](https://github.com/huggingface/optimum-quanto/issues/169#issuecomment-2323994489)? Please add a link
      to it if that's the case.
- [x] Did you run all tests locally and make sure they pass.
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

### CODE REVIEW

File: examples/vision/image-classification/pets/quantize_vit_model.py [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+    model.eval()
+    test_loss = 0
+    correct = 0
+    with torch.no_grad():
+        start = time.time()
+        for batch in test_loader:
+            data, target = batch["pixel_values"], batch["labels"]
+            data, target = data.to(device), target.to(device)
+            output = model(data).logits
+            # print("*****I am after output", output)
```
Comment (dacorvo) [inline examples/vision/image-classification/pets/quantize_vit_model.py:40]:
```suggestion
```

File: examples/vision/image-classification/pets/quantize_vit_model.py [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+def main():
+    parser = argparse.ArgumentParser(description="ViT PETS Example")
+    parser.add_argument("--model", type=str, default="super-j/vit-base-pets")
+    parser.add_argument("--device", type=str, default=None, help="The device to use for evaluation.")
+    parser.add_argument("--weights", type=str, default="int8", choices=["int4", "int8", "float8"])
+    parser.add_argument("--activations", type=str, default="int8", choices=["none", "int8", "float8"])
+    args = parser.parse_args()
+
+    # torch.manual_seed(args.seed)
+
```
Comment (dacorvo) [inline examples/vision/image-classification/pets/quantize_vit_model.py:70]:
```suggestion
```

File: examples/vision/image-classification/pets/quantize_vit_model.py [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+    dataset_kwargs = {}
+
+    if args.device is None:
+        if torch.cuda.is_available():
+            device = torch.device("cuda")
+            cuda_kwargs = {"num_workers": 1, "pin_memory": True, "shuffle": True}
+            dataset_kwargs.update(cuda_kwargs)
+        elif torch.backends.mps.is_available():
+            device = torch.device("cpu")
```
Comment (dacorvo) [inline examples/vision/image-classification/pets/quantize_vit_model.py:79]:
```suggestion
            device = torch.device("mps")
```

**Review by dacorvo — CHANGES_REQUESTED** [review]
Thank you very much for this pull-request. Can you remove the debug comments, and apply styling (`make style`) to make the CI green ?

File: examples/vision/image-classification/pets/quantize_vit_model.py (hunk 3 above)
Comment (shovan777) [reply examples/vision/image-classification/pets/quantize_vit_model.py:79]: Looks like `mps` doesn't support `float8` dtype.

Screenshot 2024-09-13 at 13 48 41

Comment (dacorvo) [reply examples/vision/image-classification/pets/quantize_vit_model.py:79]: Indeed. But your example is correct: it is just than on mac you must pass device='cpu' if you select float8.

**Comment by dacorvo** [comment]
The CPU tests are failing because of the release of pytorch 2.4.1 (I have a guard preventing to use `torch._int_mm` for versions up to 2.4.0, but it seems it is still inaccurate in 2.4.1).

File: examples/vision/image-classification/pets/quantize_vit_model.py (hunk 3 above)
Comment (shovan777) [reply examples/vision/image-classification/pets/quantize_vit_model.py]: Yes, I have checked for float8 in https://github.com/huggingface/optimum-quanto/pull/308/commits/bd0332fbaac1d7f580a1be7cadb0c972f6779536

File: examples/vision/image-classification/pets/quantize_vit_model.py (hunk 1 above)
Comment (shovan777) [reply examples/vision/image-classification/pets/quantize_vit_model.py]: removed it in https://github.com/huggingface/optimum-quanto/pull/308/commits/bd0332fbaac1d7f580a1be7cadb0c972f6779536

File: examples/vision/image-classification/pets/quantize_vit_model.py (hunk 2 above)
Comment (shovan777) [reply examples/vision/image-classification/pets/quantize_vit_model.py]: removed it in https://github.com/huggingface/optimum-quanto/pull/308/commits/bd0332fbaac1d7f580a1be7cadb0c972f6779536

**Comment by shovan777** [comment]
@dacorvo I have made the changes you requested.

**Review by dacorvo — APPROVED** [review]
Thank you for this pull-request !

**Pull request merged**